File: ExamPro/backend/scripts/import_question_bank.py
import os
import json
import glob
# pyrefly: ignore [missing-import]
from pymongo import MongoClient
from dotenv import load_dotenv

load_dotenv()

MONGO_URI = os.getenv("MONGO_URI", "mongodb://localhost:27017/exampro")
client = MongoClient(MONGO_URI)
db = client.get_database()
questions_collection = db['questions'] # You might want to change this if your collection name is different

QUESTION_BANK_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "..", "QuestionBank")

def import_question_bank():
    print(f"Starting import from {QUESTION_BANK_DIR}...")
    
    if not os.path.exists(QUESTION_BANK_DIR):
        print(f"Directory {QUESTION_BANK_DIR} does not exist. Please ensure the QuestionBank folder is created.")
        return

    # Path pattern: QuestionBank/exam/subject/topic/difficulty.json
    search_pattern = os.path.join(QUESTION_BANK_DIR, "*", "*", "*", "*.json")
    json_files = glob.glob(search_pattern)
    
    total_inserted = 0

    for file_path in json_files:
        # Extract metadata from the folder structure
        # Example: ['..', 'QuestionBank', 'jee_mains', 'physics', 'kinematics', 'easy.json']
        parts = os.path.normpath(file_path).split(os.sep)
        
        try:
            exam = parts[-4]
            subject = parts[-3]
            topic = parts[-2]
            difficulty = parts[-1].replace('.json', '').capitalize()
            
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                
            # If the json contains a single object instead of a list, wrap it in a list
            if isinstance(data, dict):
                data = [data]
                
            # Inject path metadata into each question
            for q in data:
                q['exam'] = q.get('exam', exam)
                q['subject'] = q.get('subject', subject)
                q['topic'] = q.get('topic', topic)
                q['difficulty'] = q.get('difficulty', difficulty)
                
                # Optional: format exam name for consistency
                if q['exam'] == 'jee_mains':
                    q['exam'] = 'JEE Mains'
                elif q['exam'] == 'emcet':
                    q['exam'] = 'EAMCET'

            # Insert into MongoDB
            if data:
                # Remove _id if it's there so MongoDB generates a new ObjectId, 
                # or leave it if you want to use the explicit one from JSON
                for q in data:
                    if '_id' in q and isinstance(q['_id'], dict) and '$oid' in q['_id']:
                        from bson import ObjectId
                        q['_id'] = ObjectId(q['_id']['$oid'])

                result = questions_collection.insert_many(data)
                print(f"Inserted {len(result.inserted_ids)} questions from {file_path}")
                total_inserted += len(result.inserted_ids)
                
        except Exception as e:
            print(f"Error processing file {file_path}: {e}")

    print(f"\nImport completed! Total questions inserted: {total_inserted}")

if __name__ == "__main__":
    import_question_bank()

File: ExamPro/backend/scripts/import_question_bank.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import mongoose from 'mongoose';
import dotenv from 'dotenv';
import Question from '../models/Question.js';

dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const QUESTION_BANK_DIR = path.join(__dirname, '..', '..', 'QuestionBank');
const MONGO_URI = process.env.MONGO_URI || 'mongodb://localhost:27017/exampro';

function findJsonFiles(dir) {
  let results = [];
  if (!fs.existsSync(dir)) return results;

  const list = fs.readdirSync(dir);
  list.forEach(file => {
    const fullPath = path.join(dir, file);
    const stat = fs.statSync(fullPath);
    if (stat && stat.isDirectory()) {
      results = results.concat(findJsonFiles(fullPath));
    } else if (file.endsWith('.json')) {
      results.push(fullPath);
    }
  });
  return results;
}

async function importQuestionBank() {
  console.log(`Starting import from ${QUESTION_BANK_DIR}...`);

  try {
    await mongoose.connect(MONGO_URI);
    console.log('Connected to MongoDB.');

    const jsonFiles = findJsonFiles(QUESTION_BANK_DIR);
    console.log(`Found ${jsonFiles.length} JSON file(s).`);

    let totalInserted = 0;

    for (const filePath of jsonFiles) {
      const relativePath = path.relative(QUESTION_BANK_DIR, filePath);
      const parts = relativePath.split(path.sep);

      // Example path: JeeMains/Physics/Altenate Current/easy.json
      if (parts.length < 4) continue;

      const examRaw = parts[0];
      const subjectRaw = parts[1];
      const topicRaw = parts[2];
      const difficultyFileName = parts[parts.length - 1].replace('.json', '');

      const exam = examRaw === 'JeeMains' || examRaw === 'jee_mains' ? 'JEE Mains' : examRaw;
      const subject = subjectRaw;
      const topic = topicRaw;
      const difficulty = difficultyFileName.charAt(0).toUpperCase() + difficultyFileName.slice(1).toLowerCase();

      const rawContent = fs.readFileSync(filePath, 'utf-8');
      let data = JSON.parse(rawContent);

      if (!Array.isArray(data)) {
        data = [data];
      }

      const formattedQuestions = data.map((q, idx) => {
        return {
          level: q.level || (difficulty === 'Easy' ? 1 : difficulty === 'Medium' ? 2 : 3),
          examType: q.examType || exam,
          subject: q.subject || subject,
          chapter: q.chapter || topic,
          difficulty: ['Easy', 'Medium', 'Hard'].includes(q.difficulty) ? q.difficulty : difficulty,
          questionType: q.questionType || 'MCQ',
          question: q.question,
          questionImage: q.questionImage || q.image || '',
          options: q.options || [q.optionA, q.optionB, q.optionC, q.optionD].filter(Boolean),
          correctAnswer: q.correctAnswer,
          explanation: q.explanation || '',
          explanationImage: q.explanationImage || q.solutionImage || '',
          previousYear: q.previousYear || '',
          marks: q.marks || 4,
          negativeMarks: q.negativeMarks || 1,
          status: q.status || 'Published'
        };
      });

      if (formattedQuestions.length > 0) {
        for (const q of formattedQuestions) {
          await Question.updateOne(
            { question: q.question, chapter: q.chapter, subject: q.subject },
            { $set: q },
            { upsert: true }
          );
        }
        console.log(`Successfully imported/updated ${formattedQuestions.length} questions from ${relativePath}`);
        totalInserted += formattedQuestions.length;
      }
    }

    console.log(`\nImport complete! Processed ${totalInserted} questions.`);
    process.exit(0);
  } catch (error) {
    console.error('Error importing question bank:', error);
    process.exit(1);
  }
}

importQuestionBank();
